feat: deploy command (#295)
algorandfoundation/algokit-cli: docs/features/deploy.md

- `--prod`: Skips the warning prompt for deployments to a mainnet.
- `-h, --help`: Show this message and exit.

## Functionality

AlgoKit requires certain configuration files at the root of your project to successfully perform deployment. Below is the directory layout expected by AlgoKit:

```md
.
├── ... (your project files and directories)
├── .algokit.toml # Configuration file for AlgoKit
└── .env.[{mainnet|testnet|localnet|betanet|custom}] # (OPTIONAL) Environment variables specific to deployments to a network
```

### AlgoKit Configuration File

AlgoKit allows you to define your deployment logic by adding an `.algokit.toml`, a file present at the root of your project instantiated with `algokit init`. This file should include deployment commands for the network(s) you wish to target. The syntax is as follows:

```toml
[deploy.{network_name}]
command = "deployment-command-for-{network_name}"
```

#### Example

For instance, if you're aiming to deploy to a testnet, your `.algokit.toml` file should contain:

```toml
[deploy.testnet]
command = "poetry run python -m smart_contracts deploy"
```

### Environment Files

> For each network you're deploying to, you need to have a corresponding `.env.[network_name]` file. This file should contain environment variables specific to that network.

Reviewer: This says you need the file, it’s optional right?

Author: @robdmoore see the `note` field right below it explicitly saying its optional

├── .env # (OPTIONAL) General environment variables common across all deployments
└── .env.[{mainnet|testnet|localnet|betanet|custom}] # (OPTIONAL) Environment variables specific to deployments to a network
```

> ⚠️ Please note that creating `.env` and `.env.[network_name]` files is only necessary if you're deploying to a custom network or if you want to override the default network configurations provided by AlgoKit. AlgoKit comes with predefined configurations for popular networks like `TestNet`, `MainNet`, `BetaNet`, or AlgoKit's `LocalNet`.

The logic for loading environment variables is as follows:

- If a `.env` file exists, the environment variables contained in it are loaded first.
- If a `.env.[network_name]` file exists, the environment variables in it are loaded, overriding any previously loaded values from the `.env` file for the same variables.

## AlgoKit Configuration File

AlgoKit uses a configuration file called `.algokit.toml` in the root of your project. The configuration file can be created using the `algokit init` command. This file will define the deployment commands for the various network environments that you want to target.

> Here's an example of what the `.algokit.toml` file might look like. The example below will always prompt for the deployer mnemonic secret (unless specified in env variables) for any network, except for the localnet, where it will not prompt for any secrets.

Reviewer 2: ```suggestion
Here's an example of what the `.algokit.toml` file might look like. When deploying it will prompt for the `DEPLOYER_MNEMONIC` secret unless it is already defined as an environment variable or is deploying to localnet.
```

The logic for loading environment variables is as follows:

- If a `.env` file exists, the environment variables contained in it are loaded first.
- If a `.env.[network_name]` file exists, the environment variables in it are loaded, overriding any previously loaded values from the `.env` file for the same variables.

## AlgoKit Configuration File

AlgoKit uses a configuration file called `.algokit.toml` in the root of your project. The configuration file can be created using the `algokit init` command. This file will define the deployment commands for the various network environments that you want to target.

Here's an example of what the `.algokit.toml` file might look like:

```toml
[algokit]
min_version = "v{lastest_version}"

[deploy]
command = "your command"
environment_secrets = [
  "YOUR_COMMON_SECRET",
]

[deploy.localnet]
command = "your localnet-specific command"  # Optional. If not specified, the generic command from [deploy] will be used.
environment_secrets = [

>   "YOUR_LOCALNET_SECRET",

Reviewer 2: I think it might be worth simplifying this example? While this will work, it probably isn't a typical configuration. Maybe have something similar to the template, where we have some common secrets and then override those secrets in localnet as they aren't required.

Author: Refined to include the base sample from the beaker template (deploy branch )

Example:

```sh
$ algokit deploy testnet
```

This command deploys the smart contracts to the testnet.

## Custom Project Directory

By default, the deploy command looks for the `.algokit.toml` file in the current working directory. You can specify a custom project directory using the `--project-dir` option.

Example:

```sh
$ algokit deploy testnet --project-dir="path/to/project"
```

## Custom Deploy Command

You can provide a custom deploy command using the `--custom-deploy-command` option. If this option is not provided, the deploy command will be loaded from the `.algokit.toml` file.

Example:

```sh
$ algokit deploy testnet --custom-deploy-command="your-custom-command"
```

## CI Mode

By using the `--ci` or `--non-interactive` flag, you can skip the interactive prompt for mnemonics.

This is useful in CI/CD environments where user interaction is not possible. When using this flag, you need to make sure that the mnemonics are set as environment variables.

Example:

```sh
$ algokit deploy testnet --ci
```

## Example of a Full Deployment

```sh
$ algokit deploy testnet --custom-deploy-command="your-custom-command"
```

> This example shows how to deploy smart contracts to the testnet using a custom deploy command and skipping the warning prompt. This also assumes that .algokit.toml file is present in the current working directory, and .env.testnet file is present in the current working directory and contains the required environment variables for deploying to TestNet environment.

Reviewer 2: Is the bit about a warning prompt still relevant?

Author: good catch, removed the part mentioning prompt
